### DISCUSSION TITLE
Recommendation setup management cluster(openstack)

### DISCUSSION THREAD

**Discussion by nilpntr** [clastix/cluster-api-control-plane-provider-kamaji#148] [Q&A]
Kind for a local cluster is the suggested configuration for the first management cluster you will eventually switch to a self-managed one. The Kamaji setup documentation for OpenStack states that type: LoadBalancer with service annotations is used, although it is incompatible with a local kind cluster. In order to get the initial setup utilizing LoadBalancer to function, do you suggest any other initial management cluster setup method, such as a single node virtual machine on OpenStack?

**Comment by prometherion** [comment]
Just a small update, I moved this from Issue to Discussion since it's not a bug per se, and pinging my OpenStack gurus such as @sn4psh0t @jds9090 @lentzi90, hope they can chime in.

**Reply by jds9090** [reply]
To enable LoadBalancer functionality in a KIND cluster, it's recommended to use cloud-provider-kind. This solution allows you to test cloud provider-dependent features within a local KIND environment.
- https://github.com/kubernetes-sigs/cloud-provider-kind

```
ubuntu@kamaji-dev:~$ k get no -o wide
NAME                      STATUS   ROLES           AGE    VERSION   INTERNAL-IP   EXTERNAL-IP   OS-IMAGE                         KERNEL-VERSION     CONTAINER-RUNTIME
capi-test-control-plane   Ready    control-plane   3h2m   v1.27.1   172.18.0.3    <none>        Debian GNU/Linux 11 (bullseye)   5.4.0-54-generic   containerd://1.6.21
capi-test-worker          Ready    <none>          3h2m   v1.27.1   172.18.0.2    <none>        Debian GNU/Linux 11 (bullseye)   5.4.0-54-generic   containerd://1.6.21
ubuntu@kamaji-dev:~$ k get svc
NAME                    TYPE           CLUSTER-IP      EXTERNAL-IP   PORT(S)                           AGE
kamajicontrolplane-28   LoadBalancer   10.96.217.213   172.18.0.5    30080:32278/TCP,30090:31168/TCP   13m
kubernetes              ClusterIP      10.96.0.1       <none>        443/TCP                           3h2m
ubuntu@kamaji-dev:~$ k get ep
NAME                    ENDPOINTS                                                           AGE
kamajicontrolplane-28   10.244.1.54:30090,10.244.1.55:30090,10.244.1.56:30090 + 3 more...   13m
kubernetes              172.18.0.3:6443                                                     3h2m
```

**Reply by nilpntr** [reply]
But the nodes created for the Management Cluster(in openstack) should be able to communicate with the kind load balancer right? Or can you avoid that step and just create the management cluster and move it to be self managed right away?

**Reply by prometherion** [reply]
Keep in mind that Kamaji always needs a management cluster to handle properly workloads such as Tenant Control Plane pods.

What you can do is start with kind as ignition cluster, create an OpenStack based one with control plane nodes, and finally install Kamaji there with all the required components (e.g. CPI for Load Balancer and CSI for Datastore storage).

I think kind works like a charm, but it's just an ignition environment to test the architecture.

**Reply by jds9090** [reply]
> But the nodes created for the Management Cluster(in openstack) should be able to communicate with the kind load balancer right? Or can you avoid that step and just create the management cluster and move it to be self managed right away?

There are multiple ways to enable worker nodes to access the control plane endpoint. For example, if you create a LoadBalancer as a Kubernetes service, a NodePort is also created, allowing access through NodePort. Additionally, you can configure network settings in the cloud-init setup on worker nodes to establish communication with the control plane endpoint.

I recommend using the NodePort method for enabling worker nodes to access the control plane endpoint.
